filebeat/channel/runner.go

```diff
@@ -127,27 +128,27 @@ func newCommonConfigEditor(
 		return nil, err
 	}
 
-	var indexProcessor processors.Processor
-	if !config.Index.IsEmpty() {
-		staticFields := fmtstr.FieldsForBeat(beatInfo.Beat, beatInfo.Version)
-		timestampFormat, err := fmtstr.NewTimestampFormatString(&config.Index, staticFields)
-		if err != nil {
-			return nil, err
-		}
-		indexProcessor = add_formatted_index.New(timestampFormat)
-	}
-
-	userProcessors, err := processors.New(config.Processors)
-	if err != nil {
-		return nil, err
-	}
-
 	serviceType := config.ServiceType
 	if serviceType == "" {
 		serviceType = config.Module
 	}
 
 	return func(clientCfg beat.ClientConfig) (beat.ClientConfig, error) {
+		var indexProcessor processors.Processor
+		if !config.Index.IsEmpty() {
+			staticFields := fmtstr.FieldsForBeat(beatInfo.Beat, beatInfo.Version)
+			timestampFormat, err := fmtstr.NewTimestampFormatString(&config.Index, staticFields)
+			if err != nil {
+				return clientCfg, err
+			}
+			indexProcessor = add_formatted_index.New(timestampFormat)
+		}
+
+		userProcessors, err := processors.New(config.Processors)
```

Comment:
It's not clear just reading this why this fixes the problem.

Also, does this have any other unintended side effects? If you add and then remove processors from an input in a running agent policy does it work properly?

Reply:
@cmacknz 

> It's not clear just reading this why this fixes the problem.

This function returns a factory function. The processors were defined outside of the factory function were just re-used every time the factory function is called instead of creating a new list of processor instances from the raw config.

It should not have side-effects other than creating a new list of processor instances on each input restart and configuration change, which is how it's supposed to be.

> If you add and then remove processors from an input in a running agent policy does it work properly?

This change is the opposite, it's re-initialising processors on each change, it was retaining the same list of processors before, so it's possible that what you're saying was true before this PR.

Reply:
This might affect the performance when we restart inputs very often though.